### PR TITLE
issue: 817046 Fix coverity issues

### PR DESCRIPTION
**Pull request by igor-ivanov**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/187/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/203/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/188/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/204/ for details (Mellanox internal link).

File: src/vma/sock/sockinfo_udp.cpp [hunk 1]
```diff
@@ -2169,39 +2171,41 @@ void sockinfo_udp::validate_igmpv2(flow_tuple_with_local_if& flow_key)
...
+		b_any_activity = true;
 	}
 	if (m_p_socket_stats->counters.n_rx_bytes || m_p_socket_stats->counters.n_rx_packets || m_p_socket_stats->counters.n_rx_errors || m_p_socket_stats->counters.n_rx_eagain || m_p_socket_stats->n_rx_ready_pkt_count) {
 		si_logdbg_no_funcname("Rx Offload: %d KB / %d / %d / %d [bytes/packets/eagains/errors]", m_p_socket_stats->counters.n_rx_bytes/1024, m_p_socket_stats->counters.n_rx_packets, m_p_socket_stats->counters.n_rx_eagain, m_p_socket_stats->counters.n_rx_errors);
 
-		float rx_drop_percentage = 0;
-		if (m_p_socket_stats->counters.n_rx_packets || m_p_socket_stats->n_rx_ready_pkt_count)
-			rx_drop_percentage = (float)(m_p_socket_stats->counters.n_rx_ready_byte_drop * 100) / (float)m_p_socket_stats->counters.n_rx_packets;
-		si_logdbg_no_funcname("Rx byte: max %d / dropped %d (%2.2f%%) / limit %d", m_p_socket_stats->counters.n_rx_ready_byte_max, m_p_socket_stats->counters.n_rx_ready_byte_drop, rx_drop_percentage, m_p_socket_stats->n_rx_ready_byte_limit);
+		if (m_p_socket_stats->counters.n_rx_packets) {
```
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.cpp]: why do we need this additional if() statement?

File: src/vma/util/utils.cpp [hunk 2]
```diff
@@ -935,12 +938,10 @@ bool get_bond_active_slave_name(IN const char* bond_name, OUT char* active_slave
 {
 	char active_slave_path[256] = {0};
 	sprintf(active_slave_path, BONDING_ACTIVE_SLAVE_PARAM_FILE, bond_name);
-	BULLSEYE_EXCLUDE_BLOCK_START
```
Comment (rosenbaumalex) [inline src/vma/util/utils.cpp]: why remove the BULLSEYE block out?
it is very difficult to reach failure case, so we want to block it out corm code coverage testing

File: src/vma/util/utils.h [hunk 3]
```diff
@@ -161,24 +161,24 @@ int priv_read_file(const char *path, char *buf, size_t size, vlog_levels_t log_l
  * like above 'priv_read_file' however make sure that upon success the result in buf is a null terminated string
  */
 inline int priv_safe_read_file(const char *path, char *buf, size_t size, vlog_levels_t log_level = VLOG_ERROR){
-	int ret = priv_read_file(path, buf, size-1, log_level);
-	if (0 <= ret && size) buf[ret] = '\0';
+	int ret = -1;
+	if (size > 0) {
```
Comment (rosenbaumalex) [inline src/vma/util/utils.h]: 'size' check was added in priv_read_file(), 
are you sure we need this check again here?

File: src/vma/util/utils.h [hunk 4]
```diff
@@ -161,24 +161,24 @@ int priv_read_file(const char *path, char *buf, size_t size, vlog_levels_t log_l
...
 
 /**
- * like above 'priv_try_read_file' however make sure that upon success the result in buf is a null terminated string
+ * like above however make sure that upon success the result in buf is a null terminated string and VLOG_DEBUG
  */
 inline int priv_safe_try_read_file(const char *path, char *buf, size_t size) {
-	int ret = priv_try_read_file(path, buf, size - 1);
-	if (0 <= ret && size) buf[ret] = '\0';
+	int ret = -1;
+	if (size > 0) {
```
Comment (rosenbaumalex) [inline src/vma/util/utils.h]: 'size' check repeated again. is this needed?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/190/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/206/ for details (Mellanox internal link).

File: src/vma/util/lock_wrapper.h [hunk 5]
```diff
@@ -68,7 +68,7 @@ class lock_base
 #endif
 protected:
 	tscval_t                start_lock_wait() { return 0; };
-	void                    end_lock_wait(tscval_t start_time) {start_time = 0; NOT_IN_USE(start_time);};
+	void                    end_lock_wait(tscval_t start_time) {start_time = 0; start_time = start_time;};
```
Comment (rosenbaumalex) [inline src/vma/util/lock_wrapper.h]: 1. can you explain what is wrong with NOT_IN_USE()?
2. can we remove the 'start_time = 0;'

File: src/vma/util/lock_wrapper.h [hunk 6]
```diff
@@ -269,13 +269,13 @@ class lock_mutex : public lock_base
 	inline int lock() {
 		tscval_t t = start_lock_wait();
 		int ret = pthread_mutex_lock(&m_lock);
-		lock_base::lock();
+		if (0 == ret) ret = lock_base::lock();
```
Comment (rosenbaumalex) [inline src/vma/util/lock_wrapper.h]: This might cause some performance hit.

The lock_base::lock() should be optimized out by default, since without LOCK_STATS it will statically 'return 0' and compiler remove that code.
Now there is a branch point which depends on 'ret', so i expect some penalty for each call.

File: src/vma/util/lock_wrapper.h (hunk 5 above)
Comment (igor-ivanov) [inline src/vma/util/lock_wrapper.h]: 1 .I can not explain. It is confused me too but using NOT_IN_USE() does not work when start_time=start_time does.
2. I did not investigate it.

File: src/vma/util/lock_wrapper.h (hunk 6 above)
Comment (igor-ivanov) [inline src/vma/util/lock_wrapper.h]: Of course, penalty could be.

File: src/vma/util/utils.h (hunk 3 above)
Comment (igor-ivanov) [inline src/vma/util/utils.h]: different check in fact: size>0 and (size-1 >0) inside priv_read_file().
Will see how it can be reduced

File: src/vma/util/utils.cpp (hunk 2 above)
Comment (igor-ivanov) [inline src/vma/util/utils.cpp]: It was a mistake, will update

File: src/vma/util/utils.h (hunk 3 above)
Comment (igor-ivanov) [inline src/vma/util/utils.h]: I think it is redundant.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/192/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/208/ for details (Mellanox internal link).

File: src/vma/sock/sockinfo_udp.cpp (hunk 1 above)
Comment (igor-ivanov) [inline src/vma/sock/sockinfo_udp.cpp]: we need to be sure m_p_socket_stats->counters.n_rx_packets!=0

**Comment by igor-ivanov** [comment]
@OphirMunk I would like not to squash this PR during merge

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/212/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/197/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/213/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/198/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/199/ for details (Mellanox internal link).

File: src/vma/dev/ring_simple.cpp [hunk 7]
```diff
@@ -1278,7 +1278,6 @@ bool ring_simple::is_available_qp_wr(bool b_block)
 					if (ret <= 0) {
 						ring_logdbg("failed blocking on tx cq_mgr (errno=%d %m)", errno);
 						m_lock_ring_tx_buf_wait.unlock();
-						m_lock_ring_tx.lock();
```
Comment (rosenbaumalex) [inline src/vma/dev/ring_simple.cpp]: I believe we need to return/keep the lock here.
This function is called under a taken m_lock_ring_tx.lock(), so before return we need to make sure we hold the lock again.

again below... before return we need to hold lock().

File: src/vma/sock/sockinfo_udp.cpp (hunk 1 above)
Comment (rosenbaumalex) [inline src/vma/sock/sockinfo_udp.cpp]: got it. thanks.

File: src/vma/util/lock_wrapper.h (hunk 5 above)
Comment (rosenbaumalex) [inline src/vma/util/lock_wrapper.h]: I don't feel comfortable pushing this code without further understanding (research and testing).

@igor-ivanov , can you separate this to different PR so we can further discuss. I don't want this to block everything else.

File: src/vma/util/utils.h [hunk 8]
```diff
@@ -161,24 +161,20 @@ int priv_read_file(const char *path, char *buf, size_t size, vlog_levels_t log_l
...
+	int ret = -1;
+	ret = priv_read_file(path, buf, size - 1, log_level);
+	if (0 <= ret) buf[ret] = '\0';
 	return ret;
 }
 
-/**
- * like above 'priv_read_file' only diff is that the log will go in DEBUG
- */
-inline int priv_try_read_file(const char *path, char *buf, size_t size) {
```
Comment (rosenbaumalex) [inline src/vma/util/utils.h]: is this function never called?
should we leave it for future use case?

**Comment by rosenbaumalex** [comment]
@igor-ivanov , I don't feel comfortable pushing the changes in lock_wrapper.h without further understanding (research and testing).
Can you separate this to different PR so we can further discuss. I don't want this to block everything else. I understand this will fail the new cov checks.

File: src/vma/dev/ring_simple.cpp (hunk 7 above)
Comment (igor-ivanov) [inline src/vma/dev/ring_simple.cpp]: @rosenbaumalex I looked at this code long time to understand lock/unlock mechanism in this function. I found that ring_simple::mem_buf_tx_get() and ring_simple::is_available_qp_wr() implementations are very close but there is difference in lock/unlock that was reported by coverity. I guess that current code could be reasonable but I could not understand a reason. So let keep one as is, I will suppress issue using special tag and create issue to investigate.
Do you think it is a way to go?

File: src/vma/util/utils.h (hunk 8 above)
Comment (igor-ivanov) [inline src/vma/util/utils.h]: I prefer to remove unused functions

File: src/vma/util/lock_wrapper.h (hunk 5 above)
Comment (igor-ivanov) [inline src/vma/util/lock_wrapper.h]: What risk do you see?
using 
`start_time = start_time`
instead of 
`(void) start_time`
should be safe enough. I guess that most compilers optimize both variants removing from result binary.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/219/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/232/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/220/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/233/ for details (Mellanox internal link).

**Comment by igor-ivanov** [comment]
@OphirMunk I have updated PR taking into account notes from @rosenbaumalex. Lock/unlock issues were hidden under coverity suppression tag. I think it is ready to be pushed together with https://github.com/Mellanox/libvma/pull/123. I would like to merge this PR w/o squashing.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/221/ for details (Mellanox internal link).

**Comment by rosenbaumalex** [comment]
Looks good
+1

File: src/vma/util/utils.cpp [hunk 9]
```diff
@@ -298,6 +298,9 @@ const char* iphdr_protocol_type_to_str(const int type)
 int priv_read_file(const char *path, char *buf, size_t size, vlog_levels_t log_level /*= VLOG_ERROR*/)
 {
 	int len = -1;
+	if (size <= 0) {
```
Comment (OphirMunk) [inline src/vma/util/utils.cpp]:
```
size
```
is unsigned. What is the meaning of testing if it is negative?

Comment (igor-ivanov) [inline src/vma/util/utils.cpp]: Will update with 0 check only

**Comment by igor-ivanov** [comment]
@OphirMunk I removed sanity check from priv_read_file() and move one to priv_safe_read_file() and priv_try_safe_read_file()

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/228/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/240/ for details (Mellanox internal link).
